### PR TITLE
avoid additional properties except for user defined tags

### PR DESCRIPTION
**Pull request by tejaswiMinnu**
Added a rule to check if any "additionalProperties" exists in the swagger.
If there are any "additionalProperties" having parent defined as "tags" ignore this path in this rule and consider this as exceptional case by allowing this path.

### CODE REVIEW

**Comment by konrad-jamrozik** [comment]
@tejaswiMinnu as discussed on Teams, I would appreciate if you could confirm this PR does not introduce the issue described here:
- https://github.com/Azure/azure-openapi-validator/pull/504

**Review by konrad-jamrozik — CHANGES_REQUESTED** [review]
I am requesting for [this ask](https://github.com/Azure/azure-openapi-validator/pull/510#issuecomment-1542636548) to be fulfilled before the PR is merged.

**Review by konrad-jamrozik — APPROVED** [review]
@tejaswiMinnu confirmed that the issue is gone now.

File: docs/avoid-additional-properties.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+ARM OpenAPI(swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Policy-V1-05
+
+## Output Message
+
+Definitions must not have properties named "additionalProperties" exept for user defined tags.
```
Comment (rkmanda) [inline docs/avoid-additional-properties.md:17]: >exept [](http://example.com/codeflow?start=66&length=5)

except

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import linterForRule from "./utils"
+
+let linter: Spectral
+const errorMessage = "The use of additionalProperties is not allowed except for user defined tags on tracked resources."
+beforeAll(async () => {
+  linter = await linterForRule("AvoidAdditionalProperties")
+  return linter
+})
+
+test("AvoidAdditionalProperties should find errors", () => {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:11]: >AvoidAdditionalProperties [](http://example.com/codeflow?start=6&length=25)

Can u add a test for when tags is not directly under properties and check what the behavior is?

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts (hunk 2 above)
Comment (rkmanda) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:11]: Is there any other rule that would catch the existence of "tags" other than under properties?

Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:11]: Added the test case

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      info: {
+        type: "object",
+        tags: {
+          type: "object",
+          additionalProperties: {
+            type: "object",
+            params: {
+              type: "boolean",
+            },
+          },
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:207]: This is not a good representation of swagger. In an actual swagger file, info will have properties underneath it and then the tags would show up under properties. Can u use an example that closely represents an RP swagger?

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts (hunk 2 above)
Comment (rkmanda) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:11]: tx for adding the test. I have a comment on improving the test case, pl take a look.

File: packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts (hunk 3 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:210]: I added tags under properties now to represent swagger, but either case the given expression will check for "additionalProperties" if it has parent as "tags" irrespective of where it's defined.

Comment (rkmanda) [reply packages/rulesets/src/spectral/test/avoid-additional-properties.test.ts:210]: yes I understand the logic. But lets make sure to stick to a proper swagger def in all our tests

**Review by rkmanda — APPROVED** [review]
:shipit:

**Pull request merged**